File: 09_DataStructure_Algorithm/0513_day48/04_단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        int n = words.length;
        
        // 인접리스트 초기화
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (canChange(words[i], words[j])) {
                    graph.get(i).add(j);
                    graph.get(j).add(i);
                }
            }
        }
        
        // BFS
        boolean[] visited = new boolean[n];
        Queue<int[]> queue = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            if (canChange(begin, words[i])) {
                visited[i] = true;
                queue.offer(new int[]{i, 1});
            }
        }
        
        while (!queue.isEmpty()) {
            int[] tmp = queue.poll();
            int cur = tmp[0];
            int dist = tmp[1];
            
            if (words[cur].equals(target)) return dist;
            
            for (int nxt : graph.get(cur)) {
                if (!visited[nxt]) {
                    visited[nxt] = true;
                    queue.offer(new int[]{nxt, dist + 1});
                } 
            }
        }
        
        return 0;
    }
    
    boolean canChange(String word1, String word2) {
        int diff = 0;
        
        for (int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) diff++;
            
            if (diff > 1) return false;
        }
        
        return true;
    }
}

File: 09_DataStructure_Algorithm/0513_day48/04_단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    n = len(words)
    visited = [False] * n
    queue = deque([])
    
    def canChange(word1, word2):
        count = 0
        
        for i in range(len(word1)):
            if word1[i] != word2[i]:
                count += 1
        
        return count == 1
    
    queue.append([begin, 0])
    
    while queue:
        cur, dist = queue.popleft()
        
        if cur == target:
            return dist
        
        for i in range(n):
            if not visited[i] and canChange(cur, words[i]):
                visited[i] = True
                queue.append([words[i], dist + 1])
    
    return 0